server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1950,7 +1941,7 @@ public NoOpResult noOp(final NoOp noOp) throws IOException {
     }
 
     private NoOpResult innerNoOp(final NoOp noOp) throws IOException {
-        assert readLock.isHeldByCurrentThread();
+        assert isDrainedForClose() == false;
```

Comment:
I wonder if we should move or repeat this at the end of the method? Similar for `realtimeGetUnderLock`